DSharpPlus/Clients/DiscordClient.Dispatch.cs

```diff
@@ -1775,6 +1832,175 @@ internal async Task OnVoiceServerUpdateEventAsync(string endpoint, string token,
 
         #endregion
 
+        #region Thread
+
+        internal async Task OnThreadCreateEventAsync(DiscordThreadChannel thread)
+        {
+            thread.Discord = this;
+            this.InternalGetCachedGuild(thread.GuildId)._threads[thread.Id] = thread;
+
+            await this._threadCreated.InvokeAsync(this, new ThreadCreateEventArgs { Thread = thread, Guild = thread.Guild, Parent = thread.Parent }).ConfigureAwait(false);
+        }
+
+        internal async Task OnThreadUpdateEventAsync(DiscordThreadChannel thread)
+        {
+            if (thread == null)
+                return;
+
+            DiscordThreadChannel threadOld;
+            ThreadUpdateEventArgs updateEvent;
+
+            thread.Discord = this;
+
+            var guild = thread.Guild;
+            guild.Discord = this;
+
+            var cthread = this.InternalGetCachedThread(thread.Id);
+
+            if (cthread != null) //thread is cached
+            {
+                threadOld = new DiscordThreadChannel
+                {
+                    Discord = this,
+                    GuildId = cthread.GuildId,
+                    OwnerId = cthread.OwnerId,
+                    ParentId = cthread.ParentId,
+                    Id = cthread.Id,
+                    Name = cthread.Name,
+                    Type = cthread.Type,
+                    LastMessageId = cthread.LastMessageId,
+                    MessageCount = cthread.MessageCount,
+                    MemberCount = cthread.MemberCount,
+                    ThreadMetadata = cthread.ThreadMetadata,
+                    CurrentMember = cthread.CurrentMember,
+                    //threadMembers,
+                };
+
+                updateEvent = new ThreadUpdateEventArgs
+                {
+                    ThreadAfter = thread,
+                    ThreadBefore = new Optional<DiscordThreadChannel>(threadOld),
+                    Guild = thread.Guild,
+                    Parent = thread.Parent
+                };
+            }
+            else
+            {
+                updateEvent = new ThreadUpdateEventArgs
+                {
+                    ThreadAfter = thread,
+                    Guild = thread.Guild,
+                    Parent = thread.Parent
+                };
+                guild._threads[thread.Id] = thread;
+            }
```

Comment:
I set the optional because to me it makes sense that this field is more akin to boolean value if the user is in the thread or not. That is of course, me. I'll change it not use optional within my next few commits.